File: src/main/java/ting/leetcode/java/binarytree/SubtreeOfAnotherTree.java
package ting.leetcode.java.binarytree;

public class SubtreeOfAnotherTree {
    private TreeNode target;
    private boolean isSame = false;
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        target = subRoot;
        traverse(root);
        return isSame;
    }

    private void traverse(TreeNode root) {
        if (root == null || isSame) {
            return;
        }

        if (root.val == target.val)
            isSame = isSame(root, target);

        traverse(root.left);
        traverse(root.right);
    }

    private boolean isSame(TreeNode p, TreeNode q) {
        if (p == null && q == null)
            return true;

        if (p == null || q == null)
            return false;

        if (p.val != q.val)
            return false;

        return isSame(p.left, q.left) && isSame(p.right, q.right);
    }
}

File: src/main/kotlin/ting/leetcode/tree/SubtreeOfAnotherTree.kt
package ting.leetcode.tree

class SubtreeOfAnotherTree {
    fun isSubtree(root: TreeNode?, subRoot: TreeNode?): Boolean =
        when {
            root == null && subRoot == null -> true
            root == null && subRoot != null -> false
            root != null && subRoot == null -> false
            else -> isNodeEqual(root, subRoot) || isSubtree(root?.left, subRoot) || isSubtree(root?.right, subRoot)
        }

    fun isNodeEqual(node1: TreeNode?, node2: TreeNode?): Boolean =
        when {
            node1 == null && node2 == null -> true
            node1 == null && node2 != null -> false
            node1 != null && node2 == null -> false
            else -> node1?.`val` == node2?.`val` && isNodeEqual(node1?.left, node2?.left) && isNodeEqual(node1?.right, node2?.right)
        }
}